### ISSUE TITLE
SEC-2841: AbstractUserDetailsAuthenticationProvider with a UserCache doesn't seem to honor mapped authorities

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-security#3067]
[Gregory Gerard](https://jira.spring.io/secure/ViewProfile.jspa?name=ggerard) (Migrated from [SEC-2841](https://jira.spring.io/browse/SEC-2841?redirect=false)) said:

I needed to cache authentications from Active Directory so I created my own CachingActiveDirectoryLdapAuthenticationProvider and delegate to ActiveDirectoryLdapAuthenticationProvider for the meat.

This works just fine.

However, when I added a GrantedAuthoritiesMapper to either CachingActiveDirectoryLdapAuthenticationProvider or ActiveDirectoryLdapAuthenticationProvider or both, while the mapper would get called and the first time use of a URL worked, subsequent operations would fail. When I put a filter in to see what the authorities were, the mapped versions weren't there.

My work around was to explicitly call the mapper in retrieveUser(...). This now works.

In short when caching, the mapper doesn't seem to work with cached authentications.
